animated-transformer/src/lib/transformer/gpt2.ts

```diff
@@ -0,0 +1,574 @@
+    layers: layers
+  };
+
+  if (spec.addPosEmbeddings) {
+    const posEmbedding = makeTruncNormal({
+        posId: spec.posEncodingSeqLength,
+        inputRep: spec.inputRep,
+      });
+    transformerParams.posEmbedding = posEmbedding;
+  }
+
+  if (spec.layerNorm) {
+    const layerNormParams = initLayerNormParamsWithDims(
+        spec.addLayerNormBias, {'inputRep': spec.inputRep});
+    transformerParams.layerNorm = layerNormParams
+  }
+
+  return transformerParams;
+}
+
+export type TransformerComputation = {
+  layers: BatchAttnHeadCompututation[];
+};
+
+export function computeTransformer(
+  model: {
+    config: { spec: TransformerParamSpec };
+    params: TransformerParams;
+  },
+  seqInput: GTensor<'batch' | 'pos' | 'inputRep'>,
+  generator: RandomStream
+): TransformerComputation {
+  const compute: TransformerComputation = { layers: [] };
+  let currentLayerInput = dropout(model.config.spec.computeSpec.dropoutRate, 
+    seqInput, generator.random());
+  model.params.layers.forEach((layerParams, i) => {
+    const layerCompute = computeAttnHead(
+      model.config.spec.layers[i].computeSpec,
+      layerParams,
+      currentLayerInput,
+      generator
+    );
+    compute.layers.push(layerCompute);
+    currentLayerInput = layerCompute.seqOutput;
+  });
+
+  // TODO(@aliciafmachado): Hacky way to apply layer norm after the final layer.
+  if (model.params.layerNorm) {
+    const lastBlock = compute.layers[compute.layers.length - 1];
+    const finalSeqOutput = layerNorm(model.params.layerNorm, lastBlock.seqOutput, 'inputRep', 
+        model.config.spec.computeSpec.layerNormEpsilon);
+    let finalOutput = lastBlock;
+    finalOutput.seqOutput = finalSeqOutput;
+    finalOutput.seqInput = lastBlock.seqOutput;
+    compute.layers.push(finalOutput);
+  }
+  
+  return compute;
+}
+
+/** Batch compute the loss for the last token of a transformer.
+ *
+ * params: transformer parameters.
+ * tokenEmb: embeddings for all tokens.
+ * targetTokenIdxs: a one-hot token vector for the correct token.
+ */
+export function lastTokenLogits(
+  model: {
+    params: { tokenEmbedding: GTensor<'tokenId' | 'inputRep'> };
+  },
+  computation: TransformerComputation
+): GTensor<'batch' | 'tokenId'> {
+  const lastLayer = computation.layers[computation.layers.length - 1];
+  const positionParams = lastLayer.seqOutput.unstack('pos');
+  const lastPosParams = positionParams[positionParams.length - 1];
+  const logits = lastPosParams.contract(model.params.tokenEmbedding, ['inputRep']);
+  return logits;
+}
+
+/**
+ * Returns the average per example loss for the last token prediction.
+ */
+export function lastTokenCrossEntropyLoss(
+  model: {
+    params: { tokenEmbedding: GTensor<'tokenId' | 'inputRep'> };
+  },
+  computation: TransformerComputation,
+  targetTokenIdxs: GTensor<'batch'>
+): tf.Scalar {
+  const logits = lastTokenLogits(model, computation);
+  const logProbs = logits.softmax('tokenId').log();
```

Comment:
probably want to drop that log statement now.

Reply:
Do you mean the logProbs? Why?